c/rc_lock_supply.h

```diff
@@ -0,0 +1,217 @@
+  err = ckb_checked_load_cell_data((uint8_t*)&amount, &len, 0, index, source);
+  CHECK(err);
+  if (source == CKB_SOURCE_INPUT) {
+    err = add_assign_amount(&ctx->input_amount, &amount);
+    CHECK(err);
+  } else if (source == CKB_SOURCE_OUTPUT) {
+    err = add_assign_amount(&ctx->output_amount, &amount);
+    CHECK(err);
+  } else {
+    CHECK2(false, ERROR_SUPPLY_AMOUNT);
+  }
+exit:
+  return err;
+}
+
+int compare_cells_data(size_t input_index, size_t output_index) {
+  int err = 0;
+  int err2 = 0;
+  uint8_t input_version = 0;
+  uint8_t output_version = 0;
+  uint64_t input_len = 1;
+  uint64_t output_len = 1;
+  err = ckb_load_cell_data(&input_version, &input_len, 0, input_index, CKB_SOURCE_INPUT);
+  CHECK(err);
+  err = ckb_load_cell_data(&output_version, &output_len, 0, output_index, CKB_SOURCE_OUTPUT);
+  CHECK(err);
+  CHECK2(input_len == output_len, CKB_INVALID_DATA);
+  CHECK2(input_len >= MIN_INFO_CELL_LEN, CKB_INVALID_DATA);
+  CHECK2(input_version == output_version, CKB_INVALID_DATA);
+
+  uint8_t input_buff[1024];
+  uint8_t output_buff[1024];
+  uint128_t offset = INFO_CELL_CONST_OFFSET;
+  while (offset < input_len) {
+    uint64_t input_read_len = sizeof(input_buff);
+    uint64_t output_read_len = sizeof(output_buff);
+    err = ckb_load_cell_data(input_buff, &input_read_len, offset, input_index, CKB_SOURCE_INPUT);
+    err2 = ckb_load_cell_data(output_buff, &output_read_len, offset, output_index, CKB_SOURCE_OUTPUT);
+    CHECK2(err == 0 && err == err2, CKB_INVALID_DATA);
+
+    input_read_len =
+        (input_read_len > sizeof(input_buff)) ? sizeof(input_buff) : input_read_len;
+    output_read_len =
+        (output_read_len > sizeof(input_buff)) ? sizeof(input_buff) : output_read_len;
+    CHECK2(input_read_len == output_read_len, CKB_INVALID_DATA);
+    int same = memcmp(input_buff, output_buff, input_read_len);
+    CHECK2(same == 0, CKB_INVALID_DATA);
+    offset += input_read_len;
+  }
+
+exit:
+  return err;
+}
+
+
+int iterate_by_type_script_hash(uint8_t* hash, size_t source, iterate_func_t func, SupplyContextType* ctx) {
+  int err = 0;
+  size_t i = 0;
+  uint8_t hash2[32] = {0};
+  while (1) {
+    uint64_t len = 32;
+    err = ckb_load_cell_by_field(hash2, &len, 0, i, source, CKB_CELL_FIELD_TYPE_HASH);
+    if (err == CKB_INDEX_OUT_OF_BOUND) {
+      err = 0;
+      break;
+    }
+    if (err == CKB_ITEM_MISSING) {
+      i += 1;
+      continue;
+    }
+    CHECK2(err == 0, err);
+
+    if (memcmp(hash, hash2, 32) == 0) {
+      err = func(i, source, ctx);
+      CHECK(err);
+    }
+    i += 1;
+  }
+
+exit:
+  return err;
+}
+
+int check_supply(uint8_t* cell_id) {
+  int err = 0;
+  SupplyContextType ctx = {0};
+  // locate the input info cell
+  err = iterate_by_type_script_hash(cell_id, CKB_SOURCE_INPUT, locate_info_cell, &ctx);
+  CHECK(err);
+  // locate the output info cell
+  err = iterate_by_type_script_hash(cell_id, CKB_SOURCE_OUTPUT, locate_info_cell, &ctx);
+  CHECK(err);
```

Comment:
**MUST** check `input_info_cell_count` and `output_info_cell_count` before use `input_info_cell_index` and `output_info_cell_index`.

Reply:
It's already checked in:
```C
    if (ctx->input_info_cell_count > 1) {
      return CKB_INVALID_DATA;
    } else {
      return 0;
    }
```

Reply:
fixed.